### ISSUE TITLE
Isolated service providers fail to create statements correctly and port statements to and from the main service provider.

### ISSUE THREAD

**Issue by APoole00** [espertechinc/esper#86]
**Unable to stand up statements directly in the isolated service provider:**
There are certain statements that create things that are global scope; mainly named windows, named tables and variables. However, you can not create these statements directly in the isolated provider and expect them to properly work. Then seem to only work correctly if they are created in the main provider and ported(added) to the isolated provider. The issue with this is it is completely valid to created a named window that contains a select clause. This means the statement would instantly start picking up events on the main provider and populating the window. To resolve that issue you have to lock and impact throughput to the system. 

**There is no common interface between the two providers:**
This causes an issue with have classes that stand up multiple statements as one functional unit. Ideally you would pass in the main provider or an isolated provider allowing the class to be blind to the provider it is running on. However, you have to create your own class/interface and normalize the methods across the main and isolated provider to hide it.

**Complex statement combinations need to have all statements created on the main provider and then ported(added) to the isolated provider in the  exact order they were created to even have a prayer at working correctly:**
 I have seen issues where this works on and others where even this falls apart. I originally tried to create the statement in the main and instantly port them over on statement creation. However, this never worked and even with ESPER audit annotation on it seemed like events being sent to the isolated provider were not even being picked up. I printed out the statements that were created on a test that worked on the main and compared it to the ones on the isolated and they were the exact same.

**Destroying the isolated provider should not port(add) all the statements to the main provider:** 
This should be a different method. It is inconsistent behavior between services.

**Isolated providers fail to support create schema statements:**
When creating the create schema statement and porting(adding) it to an isolated provider, I get a null pointer exception. When I try not to port(add) it to the isolated provider and leave it in the main the statements don't hang together and it fails to provide the same behavior as being created strictly on a main provider.
INFO  main StatementIsolationServiceImpl - Begin isolating statements StatementName  unit IsolatedProvider4cd22ebd-b5e6-48e0-ad2f-f29b0972778b id 0
INFO  main StatementIsolationServiceImpl - Failed isolating statements StatementName unit IsolatedProvider4cd22ebd-b5e6-48e0-ad2f-f29b0972778b id 0
ERROR main EPAdministratorIsolatedImpl - Unexpected exception taking statements: null
java.lang.NullPointerException
	at com.espertech.esper.core.service.EPAdministratorIsolatedImpl.applyFilterVersion(EPAdministratorIsolatedImpl.java:236)
	at com.espertech.esper.core.service.EPAdministratorIsolatedImpl.addStatement(EPAdministratorIsolatedImpl.java:123)
	at com.espertech.esper.core.service.EPAdministratorIsolatedImpl.addStatement(EPAdministratorIsolatedImpl.java:79)

**In summary for true Complex Event Processing, isolated service providers lack the basic functionality to be used in any practical applications:**
You should be able to create classes that stand up multiple statements, tables, schema, variables on ether a main or isolated provider. They should be able to be created directly on that provider and the each should have the same methods available through the interface. The only difference being the logic behind it. If it works on one there should be no question or hacking needed to make it work on the other.

**Comment by raguessner** [comment]
Create-schema is also global in scope.

A create-window that has a select-clause because it allows to "copy" type information from an existing event type or other expression.  It does not populate.

In the case that you have an issue where something fails and there is an exception, other than related to normal validation, please just create a GitHub issue and cut&paste a simple piece of test code to recreate (rants are not useful).

You can already use multiple EPServiceProvider instances to do exactly as you want. The isolated service provider features is meant for a more narrow set of use cases. Use cases around time-and-event-visibility as such. Example is standing-by individual statements or catching-up individual statements from historical data while other statements keep running.

**Comment by APoole00** [comment]
raguessner thanks for the quick response.

However, I am going to have to disagree on a few things here.

"A create-window that has a select-clause because it allows to "copy" type information from an existing event type or other expression. It does not populate."  

Populate - fill or be present in (a place, environment, or domain). 

16.9.5. Isolation for Named Windows and Tables gives the example of the statement @Name('A') create window MyNamedWindow#time(60) as select * from MyEvent. If this is created on the main provider any MyEvents will be present in the istream(domain) of the window and after porting(adding) to the isolated provider be present in the rstream of that same named window after 60 minutes. That is a software design flaw that make the application of the statement impartial for use in an isolated provider. You can not seed that window with historical data without locking or having it corrupted with current streaming data.

I am attempting to use isolated providers for their exact purpose according to the documentation; "An isolated service allows an application to control event visibility and the concept of time as
desired on a statement level." I have a group of 15 statements that together preform a complex event processing on time driven data task. I am trying to seed them with historical data, bring them up to current time and port them back to the main provider. It is possible that hundreds of these instances will need to be created at any given time. I am trying to minimize overhead and increase performance by not creating a whole bunch of instances.

**Comment by raguessner** [comment]
>> 16.9.5. Isolation for Named Windows and Tables gives the example 

The "create window MyNamedWindow#time(60) as select * from MyEvent" is the same as "create window MyNamedWindow#time(60) as MyEvent". This is only defining the type.

I think you are misunderstanding. Try this:
1) create named window
2) populate the named window with a view events, for example by using fire-and-forget queries
3) create a statement such as "select sum(*) from Window"
You can observe that the "sum" already reflects the events in the window. There is no need to catch up named windows from historical data.
